internal/services/template/template.go

```diff
@@ -0,0 +1,156 @@
+}
+
+func NewTemplateService(templateContract Template, ipfsClient IPFSClient, ethClient *ethclient.Client) (*Service, error) {
+	return &Service{
+		templateContract: templateContract,
+		ipfsClient:       ipfsClient,
+		sigValidator:     signature.NewValidator(ethClient),
+		cache:            make(map[string][]models.TemplateAgreement),
+	}, nil
+}
+
+// GetTemplatePermissions fetches and validates template permissions
+func (s *Service) GetTemplatePermissions(ctx context.Context, permissionTemplateID string, assetDID cloudevent.ERC721DID) (map[string]bool, error) {
+	// Check cache first
+	s.cacheMutex.RLock()
+	if cachedAgreements, exists := s.cache[permissionTemplateID]; exists {
+		s.cacheMutex.RUnlock()
+		return s.extractPermissionsFromAgreements(cachedAgreements, assetDID), nil
+	}
+	s.cacheMutex.RUnlock()
+
+	templateID, ok := big.NewInt(0).SetString(permissionTemplateID, 10)
+	if !ok {
+		return nil, fmt.Errorf("could not convert template ID string to big.Int")
+	}
+
+	opts := &bind.CallOpts{
+		Context: ctx,
+	}
+	templateData, err := s.templateContract.Templates(opts, templateID)
+	if err != nil {
+		return nil, richerrors.Error{
+			Code:        http.StatusInternalServerError,
+			Err:         fmt.Errorf("failed to get template data: %w", err),
+			ExternalMsg: "Failed to get template data",
+		}
+	}
+
+	// Fetch template document from IPFS
+	rawEvent, err := ipfsdoc.GetValidSacdDoc(ctx, templateData.Source, s.ipfsClient)
+	if err != nil {
+		return nil, err
+	}
+
+	var data models.TemplateData
+	if err := json.Unmarshal(rawEvent.Data, &data); err != nil {
+		return nil, richerrors.Error{
+			Code:        http.StatusUnauthorized,
+			Err:         fmt.Errorf("failed to parse template data: %w", err),
+			ExternalMsg: "failed to parse template data",
+		}
+	}
+
+	// Validate template owner signature
+	valid, err := s.sigValidator.ValidateSignature(ctx, rawEvent.Data, rawEvent.Signature, common.HexToAddress(data.Owner.Address))
+	if err != nil {
+		return nil, richerrors.Error{
+			Code:        http.StatusUnauthorized,
+			Err:         fmt.Errorf("failed to validate template owner signature: %w", err),
+			ExternalMsg: "failed to validate template owner signature",
+		}
+	}
+
+	if !valid {
+		return nil, richerrors.Error{
+			Code:        http.StatusForbidden,
+			ExternalMsg: "invalid template owner signature",
+		}
+	}
+
+	// Cache only the agreements
+	s.cacheMutex.Lock()
+	s.cache[permissionTemplateID] = data.Agreements
+	s.cacheMutex.Unlock()
+
+	templatePermGrants := s.extractPermissionsFromAgreements(data.Agreements, assetDID)
+	return templatePermGrants, nil
+}
+
+func (s *Service) extractPermissionsFromAgreements(agreements []models.TemplateAgreement, assetDID cloudevent.ERC721DID) map[string]bool {
+	templatePermGrants := make(map[string]bool)
+
+	for _, agreement := range agreements {
+		// TODO(lorran) agreement.Asset does not have :id, fix this if needed
+		if agreement.Asset != assetDID.String() {
```

Comment:
I think we want to check that it matches the address.
I don't think we need to error
```suggestion
		if agreement.Asset != assetDID.ContractAddress.Hex(){
```

Reply:
Ah, this was changed in this PR https://github.com/DIMO-Network/token-exchange-api/pull/83